neurow/lib/neurow/internal_api/message.ex

```diff
@@ -0,0 +1,40 @@
+defmodule Neurow.InternalApi.Message do
+  defstruct [:type, :timestamp, :payload]
+
+  def from_json(payload) when is_binary(payload) do
+    case Jason.decode(payload) do
+      {:ok, decoded_payload} -> {:ok, from_json(decoded_payload)}
+      error -> error
+    end
+  end
+
+  def from_json(payload) when is_map(payload) do
+    %Neurow.InternalApi.Message{
+      type: payload["type"],
+      timestamp: payload["timestamp"],
+      payload: payload["payload"]
+    }
+  end
+
+  def validate(message) do
```

Comment:
You can probably write this with a list of guard. May be it's cleaner.

Reply:
I tried first, but the combinatory effect make it much difficult to read.